### PR TITLE
Remove derived fields in "_source" before saving outlier events.

### LINKED ISSUE #520: When use_derived_fields is activated, it creates fields outside of the outliers dictionary.
Screenshot 2020-05-15 at 18 05 43
Screenshot 2020-05-15 at 18 05 37

### PR DESCRIPTION
**Pull request by maximilienroberti**
Should close #520 and therefore prevents the creation of fields outside of the outliers dictionary when derived_fields is activated.

### CODE REVIEW

**Comment by maximilienroberti** [comment]
@daanraman When I look at the [unit tests of outliers](https://github.com/NVISO-BE/ee-outliers/blob/6162596465e454f68b4a137ded975991c6c5f753/app/tests/unit_tests/test_analyzer_terms.py#L261), it really look like the derived fields outside of the outlier dictionary were there on purpose.
examples of unit test function name: 
```
def test_terms_use_derived_fields_in_doc(self):
def test_terms_use_derived_fields_in_outlier(self):
def test_terms_not_use_derived_fields_in_doc(self):
def test_terms_not_use_derived_fields_but_present_in_outlier(self):
```
Do you have an idea on why those unit tests exist?

**Comment by daanraman** [comment]
It's going to be a bit more complicated than just removing it like this I'm afraid. We need to investigate what the impact is on existing use cases (for example the beaconing use cases use the timestamp_day and timestamp_hour for example to create buckets - do these still work?

Also, with the approach you take, I think you will have the case where you delete a key from the dictionary that does not exist, in case you are creating new derived fields which are not present in the source document.

**Comment by maximilienroberti** [comment]
> It's going to be a bit more complicated than just removing it like this I'm afraid. We need to investigate what the impact is on existing use cases (for example the beaconing use cases use the timestamp_day and timestamp_hour for example to create buckets - do these still work?

It's working indeed. It's because I delete it at the end, just before the events are saved in Elasticsearch

> Also, with the approach you take, I think you will have the case where you delete a key from the dictionary that does not exist, in case you are creating new derived fields which are not present in the source document.

Not sure to understand completely this one but indeed they were cases where it was trying to delete fields that are not present in the source document. Let me send you a solution in the next commit.

**Comment by daanraman** [comment]
I had a look at the unit tests:

`test_terms_use_derived_fields_in_doc` is indeed explicitly looking for the derived fields in _source, but those were developed by an internship student who might not have understood that the duplication should not be there.

Let's indeed drop the fields in the non-outliers dictionary prior to writing them to the database, but in the end they should not even be added to the event in the first place. I will investigate a bit why this is done this way, you can let this one rest until I get back with my feedback.

**Comment by daanraman** [comment]
I propose you also update CHANGELOG.md with this new behavior prior to merging the PR

File: app/helpers/analyzer.py [hunk 1]
```diff
@@ -95,6 +95,8 @@ def _extract_model_settings(self):
             model_settings["should_notify"] = False
 
         model_settings["use_derived_fields"] = self.config_section.getboolean("use_derived_fields")
```
Comment (michielmeersmans) [inline app/helpers/analyzer.py:97]: There's a fallback option to set default so you could do this in one line:
self.config_section.getboolean("use_derived_fields", fallback=False)

**Pull request merged**